File: simulation/actions.py
"""Validated state-changing farm actions."""

from simulation import crop_growth, inventory
from simulation.state import InventoryLot, PlantedCrop

DEFAULT_WATERING = {
    "neglect_loss_chance_penalty_per_day": 0.05,
    "neglect_yield_penalty_per_day": 0.08,
    "max_neglect_loss_chance_bonus": 0.60,
    "max_neglect_yield_penalty": 0.80,
}
DEFAULT_FERTILIZER = {"cost": 8, "yield_bonus_pct": 0.25, "loss_chance_reduction": 0.03}


def buy_seeds(player, crop: dict, quantity: int = 1) -> bool:
    cost = crop["seed_cost"] * quantity
    if quantity <= 0 or player.money < cost:
        return False
    player.money -= cost
    player.record_expense("seeds", cost)
    player.seed_inventory[crop["id"]] = player.seed_inventory.get(crop["id"], 0) + quantity
    return True


def plant_seed(
    player, crop: dict, growth_days: int, fertilized: bool = False, fertilizer_config=None
) -> bool:
    if player.open_slots <= 0 or player.seed_inventory.get(crop["id"], 0) <= 0:
        return False
    if fertilized and player.fertilizer_inventory <= 0:
        return False
    plot_index = next((index for index, plot in enumerate(player.plots) if plot.crop is None), None)
    if plot_index is None:
        return False
    planted = PlantedCrop(
        crop["id"], player.day, growth_days, plot_index=plot_index, fertilized=fertilized
    )
    player.seed_inventory[crop["id"]] -= 1
    if fertilized:
        player.fertilizer_inventory -= 1
        player.total_fertilizer_applied += 1
    player.planted.append(planted)
    player.plots[plot_index].crop = planted
    if fertilized:
        fertilizer_config = fertilizer_config or DEFAULT_FERTILIZER
        nutrients = fertilizer_config.get(
            "nutrients_added", {"nitrogen": 0.25, "phosphorus": 0.15, "potassium": 0.15}
        )
        for name, amount in nutrients.items():
            setattr(
                player.plots[plot_index],
                name,
                min(1.0, getattr(player.plots[plot_index], name) + amount),
            )
    player.total_planted += 1
    player.crop_plant_counts[crop["id"]] = player.crop_plant_counts.get(crop["id"], 0) + 1
    return True


def water_crop(player, planted: PlantedCrop, watering_settings: dict) -> bool:
    cost = watering_settings.get("cost_per_plot", 0.0)
    if player.money < cost:
        return False
    plot = player.plots[planted.plot_index] if planted.plot_index is not None else None
    if plot is None:
        return False
    player.money -= cost
    player.record_expense("watering", cost)
    planted.last_watered_day = player.day
    planted.neglect_days = 0
    plot.moisture = min(1.0, plot.moisture + watering_settings.get("moisture_added", 0.45))
    player.total_waterings += 1
    return True


def water_farm(player, agent, crops_by_id: dict, rng, watering_settings=None) -> bool:
    """Compatibility action: water every overdue crop on one diligence roll."""
    watering_settings = watering_settings or DEFAULT_WATERING
    overdue = [
        planted
        for planted in player.planted
        if planted.neglect_days > 0
        or player.day - planted.last_watered_day
        >= crops_by_id[planted.crop_id].get("water_interval_days", 3)
    ]
    if not overdue:
        return False
    if not rng.roll_watering(agent.watering_diligence):
        for planted in overdue:
            interval = crops_by_id[planted.crop_id].get("water_interval_days", 3)
            planted.neglect_days = max(0, player.day - planted.last_watered_day - interval)
        return False
    watered = False
    for planted in overdue:
        watered = water_crop(player, planted, watering_settings) or watered
    return watered


def buy_fertilizer(player, fertilizer_config: dict, quantity: int = 1) -> bool:
    cost = fertilizer_config["cost"] * quantity
    if quantity <= 0 or player.money < cost:
        return False
    player.money -= cost
    player.record_expense("fertilizer", cost)
    player.fertilizer_inventory += quantity
    player.total_fertilizer_bought += quantity
    return True


def fertilize_crop(player, planted: PlantedCrop, fertilizer_config: dict) -> bool:
    if planted.fertilized or player.fertilizer_inventory <= 0:
        return False
    planted.fertilized = True
    player.fertilizer_inventory -= 1
    player.total_fertilizer_applied += 1
    if planted.plot_index is not None:
        plot = player.plots[planted.plot_index]
        nutrients = fertilizer_config.get(
            "nutrients_added", {"nitrogen": 0.25, "phosphorus": 0.15, "potassium": 0.15}
        )
        for name, amount in nutrients.items():
            setattr(plot, name, min(1.0, getattr(plot, name) + amount))
    return True


def harvest_mature(player, crops_by_id: dict, *args) -> bool:
    if len(args) == 1:
        rng, watering_settings, fertilizer_config = args[0], DEFAULT_WATERING, DEFAULT_FERTILIZER
    elif len(args) == 3 and hasattr(args[0], "roll_loss"):
        rng, watering_settings, fertilizer_config = args
    elif len(args) == 3:
        watering_settings, fertilizer_config, rng = args
    else:
        raise TypeError("harvest_mature expects rng or watering_settings, fertilizer_config, rng")
    watering_settings = watering_settings or DEFAULT_WATERING
    fertilizer_config = fertilizer_config or DEFAULT_FERTILIZER
    still_growing = []
    harvested_any = False
    for planted in player.planted:
        if not planted.is_mature(player.day):
            still_growing.append(planted)
            continue
        harvested_any = True
        player.total_harvest_events += 1
        crop = crops_by_id[planted.crop_id]
        plot = player.plots[planted.plot_index] if planted.plot_index is not None else None
        lost, amount = crop_growth.compute_harvest_outcome(
            planted, crop, watering_settings, fertilizer_config, rng, plot
        )
        if lost or amount <= 0:
            player.total_crops_lost += 1
            player.losses_by_cause["crop_loss"] = player.losses_by_cause.get("crop_loss", 0) + 1
        else:
            _yield_multiplier, quality_score = crop_growth.harvest_multipliers(planted, crop, plot)
            grade = crop_growth.quality_grade(quality_score)
            if grade != "rejected":
                player.inventory_lots.append(
                    InventoryLot(
                        item_id=planted.crop_id,
                        quantity=amount,
                        quality=grade,
                        produced_day=player.day,
                        shelf_life_days=crop.get("shelf_life_days", 7),
                        unit_cost=crop["seed_cost"] / amount,
                    )
                )
                player.total_harvested += amount
                player.quality_harvested[grade] = player.quality_harvested.get(grade, 0) + amount
            else:
                player.total_crops_lost += 1
                player.losses_by_cause["rejected_quality"] = (
                    player.losses_by_cause.get("rejected_quality", 0) + amount
                )
        if plot is not None:
            plot.previous_crop_family = crop.get("family", crop["id"])
            plot.crop = None
            plot.soil_health = max(0.1, plot.soil_health - 0.02)
    player.planted = still_growing
    player.rebuild_crop_inventory()
    return harvested_any


def sell_all(player, crops_by_id: dict, rng) -> tuple[float, int]:
    """Compatibility sale through a basic spot market at one rolled item price."""
    player.import_legacy_inventory(crops_by_id)
    revenue = 0.0
    sold = 0
    for crop_id in list(player.crop_inventory):
        quantity = inventory.available_quantity(player, crop_id)
        if quantity <= 0:
            continue
        crop = crops_by_id[crop_id]
        price = rng.roll_price(crop["base_price"], crop["price_variation"])
        consumed, _cost = inventory.consume(player, crop_id, quantity)
        revenue += price * consumed
        sold += consumed
    if sold:
        player.money += revenue
        player.track_peak_cash()
        player.total_revenue += revenue
        player.total_sold += sold
        player.revenue_by_channel["spot"] = player.revenue_by_channel.get("spot", 0) + revenue
    return revenue, sold


def buy_upgrade(player, upgrade: dict) -> bool:
    if upgrade["id"] in player.upgrades_owned or player.money < upgrade["cost"]:
        return False
    player.money -= upgrade["cost"]
    player.record_expense("upgrades", upgrade["cost"])
    player.upgrades_owned.add(upgrade["id"])
    player.upgrade_purchase_days[upgrade["id"]] = player.day
    effect = upgrade["effect"]
    if effect["type"] == "capacity":
        player.add_slots(effect["amount"])
    return True


def do_nothing(player) -> None:
    player.idle_days += 1

File: farm-c/src/actions.c
#include "actions.h"

#include "crop_growth.h"
#include "pyfloat.h"

/* --- simulation/actions.py:16-25 --- */

bool actions_buy_seeds(FarmState *state, const CropDef *crop, int quantity) {
    if (quantity <= 0) {
        return false;
    }
    double cost = crop->seed_cost * quantity;
    if (state->money < cost) {
        return false;
    }
    state->money -= cost;
    farm_state_record_expense(state, EXPENSE_SEEDS, cost);
    state->seed_inventory[crop->item_id] += quantity;
    return true;
}

/* --- simulation/actions.py:28-65 --- */

bool actions_plant_seed(FarmState *state, const CropDef *crop, int growth_days, bool fertilized,
                         const FertilizerConfig *fertilizer) {
    if (farm_state_open_slots(state) <= 0 || state->seed_inventory[crop->item_id] <= 0) {
        return false;
    }
    if (fertilized && state->fertilizer_inventory <= 0) {
        return false;
    }
    int plot_index = -1;
    for (size_t i = 0; i < state->plots.count; i++) {
        if (state->plots.planted_index[i] == -1) {
            plot_index = (int)i;
            break;
        }
    }
    if (plot_index == -1) {
        return false;
    }

    PlantedCrop planted = {
        .crop_item_id = crop->item_id,
        .day_planted = state->day,
        .growth_days_required = growth_days,
        .last_watered_day = state->day, /* PlantedCrop.__post_init__ */
        .neglect_days = 0,
        .fertilized = fertilized,
        .plot_index = plot_index,
        .accrued_cost = crop->seed_cost + (fertilized ? fertilizer->cost : 0.0),
    };
    if (!planted_crop_vec_push(&state->planted, planted)) {
        farm_state_mark_allocation_failed(state);
        return false;
    }
    int new_index = (int)state->planted.count - 1;

    state->seed_inventory[crop->item_id] -= 1;
    if (fertilized) {
        state->fertilizer_inventory -= 1;
        state->total_fertilizer_applied += 1;
    }
    state->plots.planted_index[plot_index] = new_index;
    if (fertilized) {
        state->plots.nitrogen[plot_index] =
            py_min(1.0, state->plots.nitrogen[plot_index] + fertilizer->nutrients_added.nitrogen);
        state->plots.phosphorus[plot_index] = py_min(
            1.0, state->plots.phosphorus[plot_index] + fertilizer->nutrients_added.phosphorus);
        state->plots.potassium[plot_index] = py_min(
            1.0, state->plots.potassium[plot_index] + fertilizer->nutrients_added.potassium);
    }
    state->total_planted += 1;
    state->crop_plant_counts[crop->item_id] += 1;
    return true;
}

/* --- simulation/actions.py:68-82 --- */

bool actions_water_crop(FarmState *state, PlantedCrop *planted, const WateringConfig *watering) {
    double cost = watering->cost_per_plot;
    if (state->money < cost) {
        return false;
    }
    if (planted->plot_index < 0 || (size_t)planted->plot_index >= state->plots.count) {
        return false;
    }
    state->money -= cost;
    farm_state_record_expense(state, EXPENSE_WATERING, cost);
    planted->accrued_cost += cost;
    planted->last_watered_day = state->day;
    planted->neglect_days = 0;
    state->plots.moisture[planted->plot_index] =
        py_min(1.0, state->plots.moisture[planted->plot_index] + watering->moisture_added);
    state->total_waterings += 1;
    return true;
}

/* --- simulation/actions.py:108-118 --- */

bool actions_buy_fertilizer(FarmState *state, const FertilizerConfig *fertilizer, int quantity) {
    if (quantity <= 0) {
        return false;
    }
    double cost = fertilizer->cost * quantity;
    if (state->money < cost) {
        return false;
    }
    state->money -= cost;
    farm_state_record_expense(state, EXPENSE_FERTILIZER, cost);
    state->fertilizer_inventory += quantity;
    state->total_fertilizer_bought += quantity;
    return true;
}

/* --- simulation/actions.py:121-135 --- */

bool actions_fertilize_crop(FarmState *state, PlantedCrop *planted,
                             const FertilizerConfig *fertilizer) {
    if (planted->fertilized || state->fertilizer_inventory <= 0) {
        return false;
    }
    planted->fertilized = true;
    state->fertilizer_inventory -= 1;
    state->total_fertilizer_applied += 1;
    planted->accrued_cost += fertilizer->cost;
    if (planted->plot_index >= 0 && (size_t)planted->plot_index < state->plots.count) {
        int plot_index = planted->plot_index;
        state->plots.nitrogen[plot_index] =
            py_min(1.0, state->plots.nitrogen[plot_index] + fertilizer->nutrients_added.nitrogen);
        state->plots.phosphorus[plot_index] = py_min(
            1.0, state->plots.phosphorus[plot_index] + fertilizer->nutrients_added.phosphorus);
        state->plots.potassium[plot_index] = py_min(
            1.0, state->plots.potassium[plot_index] + fertilizer->nutrients_added.potassium);
    }
    return true;
}

/* --- simulation/actions.py:138-206 harvest_mature (modern-path shape) --- */

bool actions_harvest_mature(FarmState *state, const ResolvedConfig *config, FarmRng *rng,
                             const WateringConfig *watering, const FertilizerConfig *fertilizer) {
    bool harvested_any = false;
    size_t mature_count = 0;
    for (size_t i = 0; i < state->planted.count; i++) {
        const PlantedCrop *planted = &state->planted.data[i];
        if (state->day - planted->day_planted >= planted->growth_days_required)
            mature_count++;
    }
    if (mature_count > 0 &&
        (mature_count > SIZE_MAX - state->inventory_lots.count ||
        !vec_reserve((void **)&state->inventory_lots.data, &state->inventory_lots.capacity,
                     state->inventory_lots.count + mature_count, sizeof(InventoryLot)))) {
        farm_state_mark_allocation_failed(state);
        return false;
    }
    /* In-place write-pointer compaction on state->planted itself, instead
     * of rebuilding into a fresh vector and swapping -- same pattern as
     * inventory.c's remove_empty_lots/processing.c's processing_complete_
     * jobs. write <= read always holds, and `planted` below is copied out
     * of state->planted.data[read] before any write, so writing through
     * data[write] (write < read) never corrupts an element not yet read. */
    size_t write = 0;

    /* Iterated in state->planted's original order, not plot order: RNG
     * draws happen below (compute_harvest_outcome) only for mature crops,
     * in this exact sequence, so reordering would silently desynchronize
     * replay for every recorded seed -- see actions.h's header comment. */
    for (size_t read = 0; read < state->planted.count; read++) {
        PlantedCrop planted = state->planted.data[read];
        bool mature = (state->day - planted.day_planted) >= planted.growth_days_required;
        if (!mature) {
            state->planted.data[write] = planted;
            if (planted.plot_index >= 0 && (size_t)planted.plot_index < state->plots.count) {
                state->plots.planted_index[planted.plot_index] = (int)write;
            }
            write++;
            continue;
        }

        harvested_any = true;
        state->total_harvest_events += 1;
        const CropDef *crop = config_find_crop(config, planted.crop_item_id);
        bool has_plot = planted.plot_index >= 0 && (size_t)planted.plot_index < state->plots.count;
        PlotState plot_row = has_plot ? plot_columns_get(&state->plots, (size_t)planted.plot_index)
                                       : (PlotState){0};
        const PlotState *plot = has_plot ? &plot_row : NULL;

        int amount;
        bool lost = crop_growth_compute_harvest_outcome(&planted, crop, watering, fertilizer, rng,
                                                          plot, &config->soil_dynamics, &amount);
        if (lost || amount <= 0) {
            state->total_crops_lost += 1;
            state->crop_loss_events += 1;
        } else {
            double yield_multiplier, quality_score;
            crop_growth_harvest_multipliers(&planted, crop, plot, fertilizer,
                                             &config->soil_dynamics, &yield_multiplier,
                                             &quality_score);
            Quality grade = crop_growth_quality_grade(quality_score);
            if (grade != QUALITY_REJECTED) {
                InventoryLot lot = {
                    .item_id = planted.crop_item_id,
                    .quantity = amount,
                    .quality = grade,
                    .produced_day = state->day,
                    .age_days = 0,
                    .shelf_life_days = crop->shelf_life_days,
                    .effective_shelf_life_days = 0, /* unset -- falls back to shelf_life_days */
                    /* Real cash spent on this planting (seed + fertilizer +
                     * watering), not seed cost alone. */
                    .unit_cost = planted.accrued_cost / amount,
                    .item_type = ITEM_CROP,
                };
                if (!inventory_lot_vec_push(&state->inventory_lots, lot)) {
                    farm_state_mark_allocation_failed(state);
                    return false;
                }
                state->total_harvested += amount;
                state->quality_harvested[grade] += amount;
            } else {
                state->total_crops_lost += 1;
                state->rejected_quality_units += amount;
            }
        }

        if (has_plot) {
            /* `crop.get("family", crop["id"])`: a family-less crop's plot
             * remembers the crop's own item id instead, so a later
             * same-family rotation check can never accidentally match a
             * different family-less crop. */
            size_t plot_index = (size_t)planted.plot_index;
            const ItemDef *item = config_find_item(config, crop->item_id);
            state->plots.previous_crop_family[plot_index] =
                crop->family != NULL ? crop->family : item->external_id;
            state->plots.planted_index[plot_index] = -1;
            state->plots.soil_health[plot_index] =
                py_max(config->soil_dynamics.min_soil_health,
                       state->plots.soil_health[plot_index] - config->soil_dynamics.harvest_soil_health_cost);
        }
    }

    state->planted.count = write;
    return harvested_any;
}

/* --- simulation/actions.py:232-242 --- */

bool actions_buy_upgrade(FarmState *state, const UpgradeDef *upgrade) {
    if (state->upgrades_owned[upgrade->id] || state->money < upgrade->cost) {
        return false;
    }
    if (upgrade->effect.type == EFFECT_CAPACITY &&
        !farm_state_add_slots(state, upgrade->effect.as.capacity)) {
        return false;
    }
    state->money -= upgrade->cost;
    farm_state_record_expense(state, EXPENSE_UPGRADES, upgrade->cost);
    state->upgrades_owned[upgrade->id] = true;
    state->upgrade_purchase_days[upgrade->id] = state->day;
    return true;
}

/* --- simulation/actions.py:245-246 --- */

void actions_do_nothing(FarmState *state) {
    state->idle_days += 1;
}
